tools/lib/gzipTileset.js

```diff
@@ -0,0 +1,110 @@
+ * @param {Boolean} [options.gzip=true] Whether to gzip or gunzip the tileset.
+ * @param {Boolean} [options.tilesOnly=false] Only gzip tiles, does not gzip tileset.json or other files.
+ * @param {Boolean} [options.verbose=false] If true prints out debug messages to the console.
+ */
+function gzipTileset(inputDirectory, outputDirectory, options) {
+    return new Promise(function(resolve, reject) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+        var gzip = defaultValue(options.gzip, true);
+        var tilesOnly = defaultValue(options.tilesOnly, false);
+        var verbose = defaultValue(options.verbose, false);
+
+        if (!defined(inputDirectory)) {
+            reject(new DeveloperError('inputDirectory is required'));
+        }
+        inputDirectory = path.normalize(inputDirectory);
+        outputDirectory = path.normalize(defaultValue(outputDirectory,
+            path.join(path.dirname(inputDirectory), path.basename(inputDirectory) + '-' + (gzip ? 'gzipped' : 'gunzipped'))));
+
+        if (verbose) {
+            console.log('Input directory: ' + inputDirectory);
+            console.log('Output directory: ' + outputDirectory);
+        }
+
+        var files = [];
+        fsExtra.walk(inputDirectory)
+            .on('data', function (item) {
+                if (!item.stats.isDirectory()) {
+                    files.push(path.relative(inputDirectory, item.path));
+                }
+            })
+            .on('end', function () {
+                if (verbose) {
+                    console.log(files.length + ' files found.');
+                    console.log((gzip ? 'Compressing' : 'Uncompressing') + ' files...');
+                }
+                Promise.map(files, function (file) {
+                    var inputFile = path.join(inputDirectory, file);
+                    var outputFile = path.join(outputDirectory, file);
+
+                    if (gzip && tilesOnly && !isTile(inputFile)) {
+                        return fsExtraCopy(inputFile, outputFile);
+                    }
+
+                    return isGzipped(inputFile)
+                        .then(function(fileIsGzipped) {
+                            if (fileIsGzipped === gzip) {
+                                return fsExtraCopy(inputFile, outputFile);
+                            }
+                            return fsExtraEnsureDir(path.dirname(outputFile))
+                                .then(function () {
+                                    var inp = fsExtra.createReadStream(inputFile);
+                                    var out = fsExtra.createWriteStream(outputFile);
+                                    var operation = gzip ? zlib.createGzip() : zlib.createGunzip();
+                                    return streamToPromise(inp.pipe(operation).pipe(out));
+                                });
+
+                        });
+                }, {concurrency: 1024})
```

Comment:
I didn't give much thought to it, I suppose keeping the number like 1024 will prevent too many promises from getting created, but still keep the operation fast. I'm not really sure what the best values are for a server or if it matters. @mramato thoughts?


and am not really sure of the implications of it for a server.